File: Engines/HRVCalculator.swift
import Foundation

struct HRVResult {
    let rmssd: Double?
    let sdnn: Double?
    let method: HRVMethod
}

enum HRVMethod {
    case rmssdFromRRIntervals
    case sdnnFromHealthKit
}

struct HRVCalculator {
    /// Compute RMSSD from raw RR intervals (in seconds).
    /// RMSSD = sqrt( (1/N) * Σ(RR[i+1] - RR[i])² )
    static func computeRMSSD(from rrIntervals: [TimeInterval]) -> Double? {
        guard rrIntervals.count > 1 else { return nil }

        // Convert timestamps to successive RR intervals
        var successiveDiffs: [Double] = []
        for i in 1..<rrIntervals.count {
            let interval = (rrIntervals[i] - rrIntervals[i - 1]) * 1000 // Convert to ms
            if interval > 200 && interval < 2000 { // Filter physiological range (30-300 BPM)
                successiveDiffs.append(interval)
            }
        }

        guard successiveDiffs.count > 1 else { return nil }

        // Compute successive differences
        var squaredDiffs: [Double] = []
        for i in 1..<successiveDiffs.count {
            let diff = successiveDiffs[i] - successiveDiffs[i - 1]
            squaredDiffs.append(diff * diff)
        }

        guard !squaredDiffs.isEmpty else { return nil }

        let meanSquaredDiff = squaredDiffs.reduce(0, +) / Double(squaredDiffs.count)
        return sqrt(meanSquaredDiff)
    }

    /// Determine best available HRV value.
    /// Prefers RMSSD from beat-to-beat data; falls back to SDNN from HealthKit.
    static func bestHRV(rrIntervals: [TimeInterval]?, sdnnValue: Double?) -> HRVResult {
        // Try RMSSD first (preferred)
        if let intervals = rrIntervals, !intervals.isEmpty {
            if let rmssd = computeRMSSD(from: intervals) {
                return HRVResult(rmssd: rmssd, sdnn: sdnnValue, method: .rmssdFromRRIntervals)
            }
        }

        // Fall back to SDNN
        if let sdnn = sdnnValue {
            return HRVResult(rmssd: nil, sdnn: sdnn, method: .sdnnFromHealthKit)
        }

        return HRVResult(rmssd: nil, sdnn: nil, method: .sdnnFromHealthKit)
    }

    /// The effective HRV value for use in recovery computation.
    /// Returns RMSSD if available, otherwise SDNN.
    static func effectiveHRV(from result: HRVResult) -> Double? {
        result.rmssd ?? result.sdnn
    }
}

File: android/core/engine/src/main/kotlin/com/apexfit/core/engine/HRVCalculator.kt
package com.apexfit.core.engine

import kotlin.math.sqrt

enum class HRVMethod {
    RMSSD_FROM_RR_INTERVALS,
    RMSSD_FROM_HEALTH_CONNECT,
    SDNN_FROM_HEALTH_CONNECT,
}

data class HRVResult(
    val rmssd: Double?,
    val sdnn: Double?,
    val method: HRVMethod,
)

object HRVCalculator {

    /**
     * Compute RMSSD from raw RR intervals (in seconds).
     * RMSSD = sqrt( (1/N) * Σ(RR[i+1] - RR[i])² )
     */
    fun computeRMSSD(rrIntervalsSeconds: List<Double>): Double? {
        if (rrIntervalsSeconds.size <= 1) return null

        // Convert timestamps to successive RR intervals in ms
        val intervals = mutableListOf<Double>()
        for (i in 1 until rrIntervalsSeconds.size) {
            val interval = (rrIntervalsSeconds[i] - rrIntervalsSeconds[i - 1]) * 1000
            if (interval in 200.0..2000.0) { // Filter physiological range (30-300 BPM)
                intervals.add(interval)
            }
        }

        if (intervals.size <= 1) return null

        // Compute successive differences
        val squaredDiffs = mutableListOf<Double>()
        for (i in 1 until intervals.size) {
            val diff = intervals[i] - intervals[i - 1]
            squaredDiffs.add(diff * diff)
        }

        if (squaredDiffs.isEmpty()) return null

        val meanSquaredDiff = squaredDiffs.sum() / squaredDiffs.size
        return sqrt(meanSquaredDiff)
    }

    /**
     * On Android with Health Connect, RMSSD is available directly.
     * This simplified version prefers the direct RMSSD value.
     */
    fun bestHRV(rmssdValue: Double? = null, sdnnValue: Double? = null): HRVResult {
        // Prefer direct RMSSD from Health Connect
        if (rmssdValue != null) {
            return HRVResult(
                rmssd = rmssdValue,
                sdnn = sdnnValue,
                method = HRVMethod.RMSSD_FROM_HEALTH_CONNECT,
            )
        }

        // Fall back to SDNN
        if (sdnnValue != null) {
            return HRVResult(
                rmssd = null,
                sdnn = sdnnValue,
                method = HRVMethod.SDNN_FROM_HEALTH_CONNECT,
            )
        }

        return HRVResult(null, null, HRVMethod.SDNN_FROM_HEALTH_CONNECT)
    }

    /**
     * The effective HRV value for use in recovery computation.
     * Returns RMSSD if available, otherwise SDNN.
     */
    fun effectiveHRV(result: HRVResult): Double? =
        result.rmssd ?: result.sdnn
}
